build/dart/rules.gni

```diff
@@ -390,3 +390,117 @@ template("application_snapshot") {
     }
   }
 }
+
+# Generates an executable from Dart code using `dart compile exe`.
+#
+# This rule is primarily intended to be used to compile test targets for use
+# by engine developers to build and run test code written in Dart in the engine
+# repo, in particular via the `et` tool. It is NOT intended for use in
+# generating production-ready user-facing binaries.
+#
+# See: //flutter/tools/engine_tool for more information on `et`.
+#
+# Arguments
+#  main_dart (required):
+#    The Dart entrypoint file.
+#
+#  package_config (optional):
+#    The packages.json file for the Dart package (if any). Required when
+#    building libraries that are part of packages with Dart dependencies.
+#    Not needed when building standalone Dart code that is not part of a
+#    package.
+#
+#  output (optional):
+#    The output executable name. By default, the target name. On Windows, an
+#    'exe' is automatically appended.
+#
+#  deps (optional):
+#    Additional dependencies. Dependencies on the frontend server and
+#    Flutter's platform.dill are included by default. This rule creates and
+#    uses a depfile, so listing all Dart sources is not necessary.
+template("dart_executable") {
+  assert(defined(invoker.main_dart), "Must specify 'main_dart'")
+
+  main_dart = invoker.main_dart
+
+  name = target_name
+
+  extra_deps = []
+  if (defined(invoker.deps)) {
+    extra_deps += invoker.deps
+  }
+  extra_inputs = [ main_dart ]
+  if (defined(invoker.inputs)) {
+    extra_inputs += invoker.inputs
+  }
+  if (defined(invoker.package_config)) {
+    package_config = rebase_path(invoker.package_config)
+    extra_inputs += [ package_config ]
+  }
+
+  ext = ""
+  if (is_win) {
+    ext = ".exe"
+  }
+  output = "$target_gen_dir/$name$ext"
+  if (defined(invoker.output)) {
+    output = invoker.output
+  }
+
+  exe_vm_args = [ "--deterministic" ]
+  if (defined(invoker.vm_args)) {
+    exe_vm_args += invoker.vm_args
+  }
+
+  abs_output = rebase_path(output)
+  exe_vm_args += [
+    "compile",
```

Comment:
Does `dart compile exe` run `pub`? If so, we'll need a way to keep it from touching the network.

Does `dart compile exe` have a flag that causes it to dump a depfile like flutter_frontend_server does? See https://github.com/flutter/engine/blob/main/build/dart/rules.gni#L71. Without that, GN won't do dependency tracking correctly. You might be able to see that in this PR by editing `et` source files, and seeing that rebuilds might not happen when you need them.

Reply:
> Does dart compile exe run pub?

That's a good question. I haven't looked at the implementation. My guess is _probably_, which we should avoid. Will take a look.

> Does dart compile exe have a flag that causes it to dump a depfile like flutter_frontend_server does?

Not one that's documented (even with `-v -v` set). This patch is still a draft experiment to test the general concept/tool integration. It should work even if we swap out the implementation with an invocation of `dart_snapshot` (wrapped with a launch script) instead, that approach does support `--depfile` (and is already implemented above). Will play with that a little.